rfcs/0000-nixpkgs-ifd.md

```diff
@@ -0,0 +1,182 @@
+
+   The benefits of IFD don't get enough attention.
+   A package in Nixpkgs is more than a derivation: there's the `meta` as I mentioned above.
+   There's also being able to read the code and (somewhat) understand what's going in reference to other derivations.
+   Finally, there's being able to `override`, `overrideAttrs`, etc. the derivation downstream.
+   IFD alone allows computed packages that follow all these norms.
+
+4. Say I want IFD that depends on other IFD?
+
+   In other words, can one import a derivation that is itself evaluated from and import from a derivation?
+   No, not without introducing another round of building and evaluating for Hydra.
+   But I don't think we need arbitrarily-deep dynamism anyways: it is a tool that should be used with care anyways, because stasis \[staticism?\] is the goodly disciplinarian that makes Nixpkgs so great.
+
+   That said, `cabal2nix` is written in Haskell, `crate2nix` is written in Rust, etc. etc.
+   We can vendor enough code to build these tools and thus bootstrap the IFD we will do.
+   Per the 3rd rule for Nixpkgs above, as long as we make the vendoring automatic and pure, this is fine, and improvement upon today.
+   Also, even if we didn't have the "one round of dynamism" restriction, we would still have the bootstrapping issue.
+
+# Alternatives
+[alternatives]: #alternatives
+
+1. Instead of evaluating Nixpkgs twice, just evaluate `allImportedDerivations` the first round.
+
+   We could do this, and would reduce total eval time, yes.
+   But, I think it would come at the cost of inciting great controversy.
+   This means users of the `enableImportFromDerivation = false` subset of Nixpkgs would still have to *wait*, for all the IFD to complete first.
+   And remember, with mass rebuilds, that could be quite some time.
+   Increasing the critical path length of *everything* we do with Nixpkgs would cause real pain in some quarters, and I don't want that to pay that as the cost of IFD.
+
+   With the plan as written, users of packages depending on IFD do have to wait slightly longer as the first eval is longer (and we wait for it before beginning to build `allImportedDerivations`).
+   But I think that is fair; we would be the "new constituency", the bottom of the pecking order, and so we should be patient so that other's workflows are not disturbed.
+
+   Longer term we could revisit this, or we could e.g. double down on automatic vendoring, committing all generated code to a second "roll-up" repo.
+   Many options between those two extremes; I rather not worry to much about it now and just take the conservative polite route proposed here to begin.
+
+2. As always, do nothing, and keep the status quo.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+Should we call it "import from Derivation", or should we give it a different name?
+`builtins.readFile <drv>` is really the same thing for our purposes, so I am sympathetic to renaming.
```

Comment:
I'm not sure I follow? To comply with the nixpkgs conventions (overrideability, `meta` sets etc.) we *need* to generate proper nix files that can be `callPackage`-ed, not derivation files.

Reply:
Oh that should be
```suggestion
`builtins.readFile <drv-output>` is really the same thing for our purposes, so I am sympathetic to renaming.
```
I agree we probably won't be doing `builtins.readFile` very often, but strictly speaking, it doesn't mean violating our conventions: there are many legitimate uses of  `builtins.fromJSON (builtins.readFile ...)` for example.

Reply:
Yes, agreed. I think IFD is the well known term for “evaluation depends on *realizing* derivations”, not sure if there's a more accurate, but catchy term to be had.

Reply:
**E**valuation-**t**ime **b**uild? The entire problem/idea is that we're building at evaluation-time, before we finish evaluation.

Reply:
Yeah I like that. My one quibble in the vein of "lazy" vs "non-strict" is that whether or not do any actually building or evaluating, the question of whether expressions use the contents of derivation outputs can be answered -- i.e. it is a static question.

Having that quibble I was still stuck with a name for 24 hours :), but now I have "build-dependent-expressions" or "output-dependent expressions" or "derivation-output-dependent-expressions" or most verbosely "content-of-derivation-outputs-dependent-expressions"

----------

On a side note, It's very sneaky with input-addressing that we can do all sorts of eval-time inspection of the output path itself without IFD. With CA derivations, output paths are not known and so we instead use opaque placeholders. This brings the store path and the content of the store path back in line, as they are both equally opaque without IFD.

It can be an annoying restriction, but I think it also makes everything easier to understand :).

Reply:
EBEB = eval build eval build